test/SolveByElim.lean

```diff
@@ -52,11 +72,62 @@ example : True := by
 example (P₁ P₂ : α → Prop) (f : ∀ (a: α), P₁ a → P₂ a → β)
     (a : α) (_ha₁ : P₁ a)
     (a' : α) (ha'₁ : P₁ a') (ha'₂ : P₂ a') : β := by
+  fail_if_success solve_by_elim (config := .noBackTracking)
+  solve_by_elim
+
+example {α : Type} {a b : α → Prop} (h₀ : b = a) (y : α) : a y = b y :=
+by
+  fail_if_success solve_by_elim (config := {symm := false})
+  solve_by_elim
+
+example (P : True → False) : 3 = 7 :=  by
+  fail_if_success solve_by_elim (config := {exfalso := false})
+  solve_by_elim
+
+-- Verifying that `solve_by_elim` acts only on the main goal.
+example (n : ℕ) : ℕ × ℕ := by
+  constructor
+  solve_by_elim
   solve_by_elim
 
--- TODO this works in mathlib3 but not here yet, for some reason.
--- example {α : Type} {a b : α → Prop} (h₀ : b = a) (y : α) : a y = b y :=
--- by solve_by_elim
+-- Verifying that `solve_by_elim*` acts on all remaining goals.
+example (n : ℕ) : ℕ × ℕ := by
+  constructor
+  solve_by_elim*
+
+-- Verifying that `solve_by_elim*` backtracks when given multiple goals.
+example (n m : ℕ) (f : ℕ → ℕ → Prop) (h : f n m) : ∃ p : ℕ × ℕ, f p.1 p.2 := by
+  fconstructor
+  fconstructor
+  solve_by_elim*
+
+-- test that metavariables created for implicit arguments don't get stuck
+example (P : ℕ → Type) (f : {n : ℕ} → P n) : P 2 × P 3 := by
+  fconstructor
+  solve_by_elim* only [f]
+
+example : 6 = 6 ∧ [7] = [7] := by
+  fconstructor
+  solve_by_elim* only [@rfl _]
+
+-- Test that `solve_by_elim*`, which works on multiple goals,
+-- successfully uses the relevant local hypotheses for each goal.
+example (f g : ℕ → Prop) : (∃ k : ℕ, f k) ∨ (∃ k : ℕ, g k) ↔ ∃ k : ℕ, f k ∨ g k := by
+  dsimp at *
+  fconstructor
+  rintro (⟨n, fn⟩ | ⟨n, gn⟩)
+  pick_goal 3
+  rintro ⟨n, hf | hg⟩
+  solve_by_elim* (config := {maxDepth := 13}) [Or.inl, Or.inr, Exists.intro]
+
+-- Test that `Config.intros` causes `solve_by_elim` to call `intro` on intermediate goals.
+example (P : Prop) : P → P := by
+  fail_if_success solve_by_elim
+  solve_by_elim (config := .intros)
+
+-- This worked in mathlib3. Why is it failing here?
+-- example (P Q R : Prop) : P ∧ Q → P ∧ Q := by
+--   solve_by_elim [And.imp, id]
```

Comment:
Weird. It gets into an infinite loop, trying to apply `id` in the wrong way.
`solve_by_elim [And.imp, @id]` works.

Reply:
Okay, I've left a comment on this test, but I'm inclined not to otherwise pursue this now.